README
PerlDeBugger (Devel-PDB) version 1.01
=====================================

(PerlDeBugger is still in development stage, not all of the planed
functions have been implemented yet)

PerlDeBugger is a Curses-based Perl debugger with most of the essential functions
such as monitoring windows for paddlist, call stack, custom watch expressions, etc.
Suitable for debugging or tracing complicated Perl applications on the spot.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Curses::UI >= 0.95

Copyright (C) 2007 by Ivan Y.C. Wong
Copyright (C) 2012 by Igor Bujna

This debugger is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Devel::PDB.

1.4   Sun Jan 12 13:40:02 CEST 2014
       - reverse for stack window
       - couple of window fixes

1.3   Sat Jan 11 20:30:02 CEST 2014
       - Better eval breakpoint (Removing from breakpoint list)
       - Repair key bind
       - In help->key on bottom can be printed duplicity in keys

1.2   Fri Jan 10 16:04:32 CEST 2014
       - Show duplicity keys in help->keys
       - Can add eval on breakpoint
       - couple of small fixes in code

1.1   Thu Jan  2 00:01:56 CEST 2014
       - Change path for setting file

1.00  Tue Jul  9 09:36:56 CEST 2013
	- Change list - can find any object in watches etc
        - new key CTRL+I for refreshing every window content
        - Add Data::Dumper::Sortkeys for watches

0.99  Mon Jul 23 15:32:23 CEST 2012
	- Eval code
	- Find perl files
	- Special variables
	- Enviroment change
	- etc ....

0.9   Tue Jul 17 15:28:53 CEST 2012
	- Stack window
	- Breakpoints listing
	- etc ....

0.07  Tue Jul 10 01:47:58 PDT 2012
	- Save|load config files(watches+breakpoints)
	- Viewing STDERR+STDOUT files
	- Keybindings
	- Exit menu
	- etc.....

0.06  Fri Jun 22 00:52:42 2007
	- fixed Watch (value was not correctly shown for newly added expression)

0.01  Sat Mar 10 01:45:08 2007
	- original version; created by h2xs 1.23 with options
		-X -n Devel::PDB -b 5.6.1



lib/Devel/PDB/Source.pm
# vi: set autoindent shiftwidth=4 tabstop=8 softtabstop=4 expandtab:
package Devel::PDB::Source;
use strict;
use warnings;

use base qw(Class::Accessor::Faster);

use Carp;

__PACKAGE__->mk_accessors(qw(
      filename
      lines
      breaks
      current_line
      scr_x
      scr_y
      cur_x
      cur_y
      view
));

sub new {
    my $class  = shift;
    my %params = @_;

    my $this = $class->SUPER::new({
            filename     => undef,
            lines        => undef,
            breaks       => undef,
            current_line => -1,
            scr_x        => 0,
            scr_y        => 0,
            cur_x        => 0,
            cur_y        => 0,
            @_,
        });

    croak 'undefined source?'           unless defined $this->lines;
    croak 'undefined breakpoint table?' unless defined $this->breaks;

    return $this;
}

sub current_line {
    my $this = shift;
    my $ret  = $this->_current_line_accessor(@_);

    if (@_) {
        $this->cur_x(0);
        $this->cur_y(@_);

        my $view = $this->view;
        $view->scroll_to_line if @_ && defined $view;
    }

    return $ret;
}

sub cur_y {
    my $this = shift;

    if (@_) {
        my ($line) = @_;
        my $line_cnt = scalar @{$this->lines};

        $line = 0             if $line < 0;
        $line = $line_cnt - 1 if $line >= $line_cnt;

        my $ret = $this->_cur_y_accessor($line);

        return $ret;
    }

    return $this->_cur_y_accessor;
}

sub toogle_break_cond {
    my ($this, $line, $str) = @_;
    my $breaks = $this->breaks;

    my ($stop, $action) = split(/\0/, $breaks->{$line});

    my $x = "";
    for (1 .. ($ENV{COLS} - 20)) {
        $x .= "-";
    }

    my $cond = $Curses::UI::rootobject->question(
        -question => $str . "\n$x",
        %DB::def_style,
        -title  => "Eval in breakpoint",
        -answer => $action,
    );
    return 0 if (!$cond || !length($cond));
    my $str = $cond;
    $str =~ s/ //g;
    return 0 unless length($str);
    return "1\0" . $cond;
}

sub ret_line_number {
    my $this = shift;

    my $line     = $this->cur_y;
    my $line_cnt = scalar @{$this->lines};

    ++$line while (${$this->lines}[$line] == 0 && $line_cnt > $line);
    return 0 if $line_cnt <= $line;

    return $line;
}

sub ret_line_breakpoint {
    my $this   = shift;
    my $line   = $this->ret_line_number() || return;
    my $breaks = $this->breaks;

    return $breaks->{$line};
}

sub toggle_break {
    my ($this, $code, $r_str) = @_;

    my $line   = $this->ret_line_number() || return 0;
    my $breaks = $this->breaks;
    my $ret    = 0;
    my $view   = $this->view;
    my $text   = ref($r_str) ? "Problem : $$r_str" : "In line $line";

    if ($breaks->{$line}) {
        if ($code) {
            my $result = $this->toogle_break_cond($line, $text);
            return $this->toggle_break() unless ($result);
            $ret = $breaks->{$line} = $result;
        } else {
            $breaks->{$line} = 0;
            delete $breaks->{$line};
        }
    } else {
        my $result = $code ? $this->toogle_break_cond($line, $text) : 1;
        $ret = $breaks->{$line} = $result;
    }

    if (defined $view) {
        if ($ret && $this->cur_y != $line) {
            $view->cursor_down(undef, $line - $this->cur_y);
        } else {
            $view->intellidraw;
        }
    }

    return $ret;
}

1;


lib/Devel/PDB/NamedListbox.pm
# vi: set autoindent shiftwidth=4 tabstop=8 softtabstop=4 expandtab:
package Devel::PDB::NamedListbox;
use strict;
use warnings;

use Curses;
use Curses::UI::Common;
use Curses::UI::Listbox;
use Devel::PDB::Dialog::Message;

use vars qw(
  $VERSION
  @ISA
  );

@ISA = qw(
  Curses::UI::Listbox
  Curses::UI::Common
  );

$VERSION = '1.1';

my $var_name_spaces = 20;

sub new {
    my $class = shift;
    my $this  = $class->SUPER::new(@_);

    $this->set_binding(\&delete_item, KEY_DC) unless $this->{-readonly};
    $this->set_binding(\&show_item, KEY_ENTER);

    $this;
}

sub delete_item {
    my $this       = shift;
    my $id         = $this->get_active_id;
    my $named_list = $this->{-named_list};

    splice @$named_list, $id, 1;
    $this->update;
}

sub show_item {
    my $this = shift;
    my $id   = $this->get_active_id;
    my $item = $this->{-named_list}->[$id];

    Devel::PDB::Dialog::Message->run(-title => $item->{name}, -message => $item->{long_value}, %DB::def_style);
}

sub named_list {
    my ($this, $list) = @_;

    $this->{-named_list} = $list if defined $list;

    $this->{-named_list};
}

sub update {
    my ($this, $refresh) = @_;
    my $list = $this->{-named_list};
    my @display;

    if ($this->{-sort_key}) {

        # Must sort array like this, other methods not worked properly
        my @a = sort { $a->{name} cmp $b->{name} } @$list;
        @$list = @a;
    }

    foreach my $item (@$list) {
        my $name = $item->{name};
        $name = substr($name, 0, $var_name_spaces - 2) . '..' if length $name > $var_name_spaces;
        push @display, $name . ' ' x ($var_name_spaces + 1 - length($name)) . $item->{value};
    }
    $this->{-values} = \@display;

    if ($refresh) {
        $this->clear_selection;
        $this->option_first if @display;
    }

    $this->schedule_draw(1);
}

1;


lib/Devel/PDB/Dialog/Message.pm
# vi: set autoindent shiftwidth=4 tabstop=8 softtabstop=4 expandtab:
package Devel::PDB::Dialog::Message;
use strict;
use warnings;

use Curses;
use Curses::UI;
use Curses::UI::Window;
use Curses::UI::Common;

use vars qw(
  $VERSION
  @ISA
  );

@ISA = qw(
  Curses::UI::Window
  Curses::UI::Common
  );

$VERSION = '1.1';

sub new () {
    my $class = shift;

    my %userargs = @_;
    keys_to_lowercase(\%userargs);

    my %args = (
        -border  => 1,
        -message => '',    # The message to show
        -ipad    => 1,
        -fg      => -1,
        -bg      => -1,

        %userargs,

        -titleinverse => 1,
        -centered     => 1,
    );

    # Create a new object, but remember the current
    # screen_too_small setting. The width needed for the
    # buttons can only be computed in the second run
    # of focus() and we do not want the first run to
    # set screen_too_small to a true value because
    # of this.
    #
    my $remember = $Curses::UI::screen_too_small;
    my $this     = $class->SUPER::new(%args);

    my $viewer = $this->add(
        'message', 'TextViewer',
        -border     => 1,
        -vscrollbar => 1,
        -wrapping   => 1,
        -padbottom  => 2,
        -text       => $this->{-message},
        -bg         => $this->{-bg},
        -fg         => $this->{-fg},
        -bbg        => $this->{-bg},
        -bfg        => $this->{-fg},
        -focusable  => 0,
    );

    # Create a hash with arguments that may be passed to
    # the Buttonbox class.
    my %buttonargs = (-buttonalignment => 'right',);
    foreach my $arg (qw(-buttons -selected -buttonalignment)) {
        $buttonargs{$arg} = $this->{$arg}
          if exists $this->{$arg};
    }
    my $b = $this->add(
        'buttons', 'Buttonbox',
        -y  => -1,
        -bg => $this->{-bg},
        -fg => $this->{-fg},

        %buttonargs
    );

    # Let the window in which the buttons are loose focus
    # if a button is pressed.
    $b->set_routine(
        'press-button',
        sub {
            my $this   = shift;
            my $parent = $this->parent;
            $parent->loose_focus();
        });

    $this->set_binding(sub { $viewer->cursor_left;     $viewer->intellidraw }, KEY_LEFT);
    $this->set_binding(sub { $viewer->cursor_right;    $viewer->intellidraw }, KEY_RIGHT);
    $this->set_binding(sub { $viewer->cursor_up;       $viewer->intellidraw }, KEY_UP);
    $this->set_binding(sub { $viewer->cursor_down;     $viewer->intellidraw }, KEY_DOWN);
    $this->set_binding(sub { $viewer->cursor_pageup;   $viewer->intellidraw }, KEY_PPAGE);
    $this->set_binding(sub { $viewer->cursor_pagedown; $viewer->intellidraw }, KEY_NPAGE);
    $this->set_binding(sub { $viewer->cursor_to_home;  $viewer->intellidraw }, KEY_HOME);
    $this->set_binding(sub { $viewer->cursor_to_end;   $viewer->intellidraw }, KEY_END);
    $this->set_binding(sub { $viewer->search_forward;  $viewer->intellidraw }, "/");
    $this->set_binding(sub { $viewer->search_backward; $viewer->intellidraw }, "?");

    # Restore screen_too_small (see above) and
    # start the second layout pass.
    $Curses::UI::screen_too_small = $remember;
    $this->layout;

    # Set the initial focus to the buttons.
    $b->focus;

    return bless $this, $class;
}

# TODO delete_curses_windows
sub layout() {
    my $this = shift;
    return $this if $Curses::UI::screen_too_small;

    # The maximum available space on the screen.
    my $avail_width  = $ENV{COLS};
    my $avail_height = $ENV{LINES};

    # Compute the maximum available space for the message.

    $this->process_padding;

    my $avail_textwidth = $avail_width;
    $avail_textwidth -= 2;                                          # border for the textviewer
    $avail_textwidth -= 2 if $this->{-border};
    $avail_textwidth -= $this->{-ipadleft} - $this->{-ipadright};

    my $avail_textheight = $avail_height;
    $avail_textheight -= 2;                                          # border for the textviewer
    $avail_textheight -= 2;                                          # empty line and line of buttons
    $avail_textheight -= 2 if $this->{-border};
    $avail_textheight -= $this->{-ipadtop} - $this->{-ipadbottom};

    # Break up the message in separate lines if neccessary.
    my @lines = ();
    foreach (split(/\n/, $this->{-message})) {
        push @lines, @{text_wrap($_, $avail_textwidth)};
    }

    # Compute the longest line in the message.
    my $longest_line = 0;
    foreach (@lines) {
        $longest_line = length($_)
          if (length($_) > $longest_line);
    }

    # Compute the width of the buttons (if the buttons
    # object is available. This is not the case just after
    # new() calls SUPER::new()).
    my $buttons      = $this->getobj('buttons');
    my $button_width = 0;
    if (defined $buttons) {
        $button_width = $buttons->compute_buttonwidth;
    }

    # Decide what is the longest line.
    $longest_line = $button_width if $longest_line < $button_width;

    # Check if there is enough space to show the widget.
    if ($avail_textheight < 1 or $avail_textwidth < $longest_line) {
        $Curses::UI::screen_too_small = 1;
        return $this;
    }

    # Compute the size of the widget.

    my $w = $longest_line;
    $w += 2;                                          # border of textviewer
    $w += 2;                                          # extra width for preventing wrapping of text
    $w += 2 if $this->{-border};
    $w += $this->{-ipadleft} + $this->{-ipadright};

    my $h = @lines;
    $h += 2;                                          # empty line + line of buttons
    $h += 2;                                          # border of textviewer
    $h += 2 if $this->{-border};
    $h += $this->{-ipadtop} + $this->{-ipadbottom};

    $this->{-width}  = $w;
    $this->{-height} = $h;

    $this->SUPER::layout;

    return $this;
}

sub get() {
    my $this = shift;
    $this->getobj('buttons')->get;
}

sub run {
    $Curses::UI::rootobject->tempdialog(@_);
}

1;


lib/Devel/PDB/Dialog/FileBrowser.pm
# vi: set autoindent shiftwidth=4 tabstop=8 softtabstop=4 expandtab:
package Devel::PDB::Dialog::FileBrowser;
use strict;
use warnings;

use Curses;
use Curses::UI::Window;
use Curses::UI::Common;

use vars qw(
  $VERSION
  @ISA
  );

@ISA = qw(
  Curses::UI::Window
  Curses::UI::Common
  );

$VERSION = '1.1';

my $file_label   = 'File:';
my $filter_label = 'Filter:';
my $dialog_title = 'File(s)';

sub new {
    my $class    = shift;
    my %userargs = @_;

    keys_to_lowercase(\%userargs);

    my %args = (
        -title => $dialog_title,
        -bg    => -1,
        -fg    => -1,

        %userargs,

        -border       => 1,
        -centered     => 1,
        -titleinverse => 0,
        -ipad         => 1,
    );

    my $this = $class->SUPER::new(%args);
    $this->root->error('No files array is given') if !defined $this->{-files};
    $this->layout();

    my $filebrowser = $this->add(
        'filebrowser', 'Listbox',
        -y           => 0,
        -border      => 1,
        -width       => $this->canvaswidth,
        -padbottom   => 4,
        -values      => [],
        -vscrollbar  => 1,
        -bg          => $this->{-bg},
        -fg          => $this->{-fg},
        -bbg         => $this->{-bg},
        -bfg         => $this->{-fg},
        -onselchange => \&on_file_active,
    );

    $filebrowser->set_routine('option-select', \&on_file_sel);

    $this->add(
        'filterlabel', 'Label',
        -x    => 1,
        -y    => $this->canvasheight - 4,
        -text => $filter_label,
        -bg   => $this->{-bg},
        -fg   => $this->{-fg},
    );

    my $filter = $this->add(
        'filter', 'TextEntry',
        -x         => 10,
        -y         => $this->canvasheight - 4,
        -text      => '',
        -width     => $this->canvaswidth - 11,
        -showlines => 1,
        -bg        => $this->{-bg},
        -fg        => $this->{-fg},
    );

    $this->add(
        'filelabel', 'Label',
        -x    => 1,
        -y    => $this->canvasheight - 3,
        -text => $file_label,
        -bg   => $this->{-bg},
        -fg   => $this->{-fg},
    );

    $this->add(
        'file', 'Label',
        -x      => 10,
        -y      => $this->canvasheight - 3,
        -width  => $this->canvaswidth - 11,
        -height => 2,
        -text   => '',
        -bg     => $this->{-bg},
        -fg     => $this->{-fg},
    );

    $this->set_binding(
        sub {
            ($this->getfocusobj == $filter ? $filebrowser : $filter)->focus;
        },
        CUI_TAB
    );

    $filter->onChange(
        sub {
            $this->refresh_list;
        });

    $this->set_binding(
        sub {
            shift->loose_focus;
        },
        CUI_ESCAPE,
        KEY_F(10));

    my @a_help = ("ESC,F10 - Exit", "Return - Choose", "Tab - switch to filter");
    if ($this->{-its_breakpoints}) {
        push(@a_help, "Del,F2 - Delete breakpoint");
        $this->set_binding(
            sub {
                my $this    = shift;
                my $browser = $this->getobj('filebrowser');
                my $id      = $browser->get_active_value();
                my $i       = 0;
                foreach my $f (@{$this->{-files}}) {
                    if ($id eq $f) {
                        splice(@{$this->{-files}}, $i, 1);
                        last;
                    }
                    $i++;
                }
                $this->refresh_list;
            },
            KEY_DC,
            KEY_F(2));
    }

    $this->add(
        'helplabel', 'Label',
        -x       => 1,
        -y       => -1,
        -reverse => 1,
        -text    => join("  |  ", @a_help));

    $this->layout();
    $this->refresh_list;

    return bless $this, $class;
}

sub layout {
    my $this = shift;

    $this->{-width}  = $this->root->{-width} - 10;
    $this->{-height} = $this->root->{-height} - 10;

    $this->SUPER::layout() or return;

    return $this;
}

sub refresh_list {
    my $this = shift;

    my $file    = $this->getobj('file');
    my $browser = $this->getobj('filebrowser');
    my $filter  = $this->getobj('filter');

    my @visible_files = sort @{$this->{-files}};
    my $regexp        = $filter->text;
    eval { @visible_files = grep /$regexp/i, @visible_files if $regexp; };

    $browser->values(\@visible_files);
    $browser->{-ypos}     = 0;
    $browser->{-selected} = undef;
    $browser->layout_content->draw(1);

    return $this;
}

sub on_file_active {
    my $browser = shift;
    my $this    = $browser->parent;
    my $file    = $this->getobj('file');
    my $active  = $browser->get_active_value;
    my $width   = $file->{-width};

    my $show = '';
    if (defined $active) {
        my $ptr = 0;
        my $len = length($active);
        for ($ptr = 0; $ptr < $len; $ptr += $width) {
            $show .= substr($active, $ptr, $width) . "\n";
        }
        $show .= substr $active, $ptr if $ptr < $len;
    }

    $file->text($show);
}

sub on_file_sel {
    my $browser = shift;
    my $this    = $browser->parent;

    $this->{-file} = $browser->get_active_value;
    $this->loose_focus;
}

sub draw(;$) {
    my $this = shift;
    my $no_doupdate = shift || 0;

    # Draw Window
    $this->SUPER::draw(1) or return $this;

    $this->{-canvasscr}->noutrefresh();
    doupdate() unless $no_doupdate;

    return $this;
}

sub get() {
    return shift->{-file};
}

1;


lib/Devel/PDB/SourceView.pm
# vi: set autoindent shiftwidth=4 tabstop=8 softtabstop=4 expandtab:
package Devel::PDB::SourceView;

use strict;
use warnings;

use Curses;
use Curses::UI::Widget;
use Curses::UI::Common;

use Devel::PDB::Source;

use vars qw(
  $VERSION
  @ISA
  );

$VERSION = '1.1';

@ISA = qw(
  Curses::UI::Widget
  );

sub new () {
    my ($class, %userargs) = @_;

    keys_to_lowercase(\%userargs);

    my %args = (
        -parent => undef,    # the parent window
        -width  => undef,    # the width of the label
        -height => undef,    # the height of the label
        -x      => 0,        # the hor. pos. rel. to the parent
        -y      => 0,        # the vert. pos. rel. to the parent
        -bg     => -1,
        -fg     => -1,
        -source => undef,

        %userargs,

        -routines => {
            'cursor-up'    => \&cursor_up,
            'cursor-down'  => \&cursor_down,
            'cursor-ppage' => \&cursor_pageup,
            'cursor-npage' => \&cursor_pagedown,
            'search'       => \&search,
            'search-next'  => \&search_next,
            'search-prev'  => \&search_prev,
            'goto'         => \&goto,
            'cursor-home'  => \&cursor_home,
            'cursor-end'   => \&cursor_end,
        },
        -bindings => {
            KEY_UP()    => 'cursor-up',
            'k'         => 'cursor-up',
            KEY_DOWN()  => 'cursor-down',
            'j'         => 'cursor-down',
            KEY_PPAGE() => 'cursor-ppage',
            "\cB"       => 'cursor-ppage',
            KEY_NPAGE() => 'cursor-npage',
            "\cF"       => 'cursor-npage',
            '/'         => 'search',
            'n'         => 'search-next',
            'N'         => 'search-prev',
            "\cG"       => 'goto',
            KEY_HOME()  => 'cursor-home',
            KEY_END()   => 'cursor-end',
        },

        -nocursor => 0,
    );

    # Create the widget.
    my $this = $class->SUPER::new(%args);

    $this->layout();

    return $this;
}

sub layout() {
    my $this = shift;
    $this->SUPER::layout or return;

    return $this;
}

sub source {
    my $this = shift;

    if (@_) {
        my $source = shift;
        $this->{-source}->view(undef) if $this->{-source};
        $this->{-source} = $source;
        $this->{-source}->view($this);
        return $this;
    }

    return $this->{-source};
}

sub scroll_to_line {
    my $this         = shift;
    my $source       = $this->{-source};
    my $current_line = $source->current_line;
    my $y1           = $source->scr_y;
    my $y2           = $y1 + $this->canvasheight;

    if ($current_line < $y1 || $current_line >= $y2) {
        $y1 = $current_line - ($this->canvasheight >> 1);
        $y1 = 0 if $y1 < 0;
        $source->scr_y($y1);
    }
}

sub draw(;$) {
    my $this      = shift;
    my $no_update = shift || 0;
    my $source    = $this->{-source};

    $this->{-title} = $source->filename . ':' . ($source->cur_y + 1);

    # Draw the widget.
    $this->SUPER::draw(1) or return $this;

    my $canvas = $this->{-canvasscr};

    # Clear all attributes.
    $canvas->attrset(A_NORMAL);

    # Let there be color
    my $color;

    if ($Curses::UI::color_support) {
        $color = COLOR_PAIR($Curses::UI::color_object->get_color_pair($this->{-fg}, $this->{-bg}));
        $canvas->attron($color);
    }

    my $current_line = $source->current_line;
    my $lines        = $source->lines;
    my $breaks       = $source->breaks;
    my $cwidth       = $this->canvaswidth - 2;
    my $cheight      = $this->canvasheight;

    for (my ($n, $y) = ($source->scr_y, 0); $n < @$lines && $y < $cheight; ++$n, ++$y) {
        my $line = $lines->[$n] || '#';
        my $reverse = $current_line == $n;

        chomp $line;

        # Clip it if it is too long.
        $line = substr($line, 0, $cwidth) if length($line) > $cwidth;

        if ($reverse) {
            $canvas->attron(A_REVERSE);
            $canvas->addstr($y, 0, ' ' x $cwidth);
        }

        if ($breaks->{$n}) {
            $canvas->attron(
                COLOR_PAIR($Curses::UI::color_object->get_color_pair(($breaks->{$n} =~ /\0/) ? 'black' : 'red', $this->{-bg})))
              if $color;
            $canvas->addch($y, 0, '*');
            $canvas->attron($color) if $color;
        }
        $canvas->addstr($y, 2, $line);

        $canvas->attroff(A_REVERSE) if $reverse;
    }

    $canvas->move($source->cur_y - $source->scr_y, $source->cur_x - $source->scr_x + 2);

    $canvas->noutrefresh;
    doupdate() unless $no_update;

    return $this;
}

sub scroll_to_cursor {
    my ($this) = @_;
    my $source = $this->source;
    my $cur_y  = $source->cur_y;
    my $scr_y  = $source->scr_y;
    my $height = $this->canvasheight;

    $source->scr_y($cur_y - $height + 1) if $cur_y >= $scr_y + $height;
    $source->scr_y($cur_y) if $cur_y < $scr_y;

    $this->intellidraw;
}

sub cursor_up(;$) {
    my $this = shift;
    shift;    # stub for bindings handling.
    my $amount = shift || 1;
    my $source = $this->source;

    $source->cur_y($source->cur_y - $amount);
    $this->scroll_to_cursor;

    return $this;
}

sub cursor_down(;$) {
    my $this = shift;
    shift;    # stub for bindings handling.
    my $amount = shift || 1;
    my $source = $this->source;

    $source->cur_y($source->cur_y + $amount);
    $this->scroll_to_cursor;

    return $this;
}

sub cursor_pageup(;$) {
    my $this = shift;

    $this->cursor_up(undef, $this->canvasheight - 1);

    return $this;
}

sub cursor_pagedown(;$) {
    my $this = shift;

    $this->cursor_down(undef, $this->canvasheight - 1);

    return $this;
}

sub real_search {
    my ($this, $dir, $regex) = @_;

    $regex = $Curses::UI::rootobject->question(-question => 'Please enter a RegEx to search for', %DB::def_style) if !$regex;

    if ($regex) {
        my $source = $this->source;
        my $lines  = $source->lines;
        my $cnt    = @$lines;
        my $i      = $source->cur_y + $dir;
        for (; $i >= 0 && $i < $cnt; $i += $dir) {
            if ($lines->[$i] =~ /$regex/i) {
                $source->cur_y($i);
                $this->scroll_to_cursor;
                last;
            }
        }
        $this->{-lastsearch} = $regex;
    }
}

sub search {
    my $this = shift;
    $this->real_search(1);
}

sub search_next {
    my $this = shift;
    $this->real_search(1, $this->{-lastsearch});
}

sub search_prev {
    my $this = shift;
    $this->real_search(-1, $this->{-lastsearch});
}

sub goto {
    my ($this, $line) = @_;

    $line = undef if (length($line) == 1 && ord($line) < 32);
    $line = $Curses::UI::rootobject->question(-question => 'Destination line number', %DB::def_style) unless (defined($line));
    $line = int $line if defined $line;
    if ($line > 0) {
        my $source = $this->source;
        my $cnt    = @{$source->lines};

        $line = $cnt if $line > $cnt;
        $source->cur_y($line - 1);
        $this->scroll_to_cursor;
    }
}

sub cursor_home {
    my $this = shift;
    $this->goto(1);
}

sub cursor_end {
    my $this   = shift;
    my $source = $this->source;
    $this->goto(scalar(@{$source->lines}));
}

1;
